src/backend/access/aocs/aocssegfiles.c

```diff
@@ -100,9 +100,11 @@ InsertInitialAOCSFileSegInfo(Relation prel, int32 segno, int32 nvp)
 
 	segrel = heap_open(prel->rd_appendonly->segrelid, RowExclusiveLock);
 
-	InsertFastSequenceEntry(prel->rd_appendonly->segrelid,
-							(int64) segno,
-							0);
+	/*
+	 * Create a new entry in gp_fastsequence or increment
+	 * an existing one (it can exist after transaction rollback).
+	 */
+	GetFastSequences(prel->rd_appendonly->segrelid, (int64) segno, 0, 0);
```

Comment:
Both functions `InsertInitialAOCSFileSegInfo` and `GetFastSequences` uses `insert_or_update_fastsequence` function. This function was called with `newLastSequence=0` before proposed changes. But we pass -1 (sum of the third and fourth arguments of `GetFastSequences` - 1) after this changes. It looks strange.

Reply:
@Stolb27 your remark is correct - `InsertInitialAOCSFileSegInfo()` produces `last_sequence = -1` when we initialize a new segment without any dependent tuples in `gp_fastsequence`. This problem doesn't affect segments as the only place where we call `InsertInitialAOCSFileSegInfo()` there is an `aocs_insert_init()` function where we call [GetFastSequences()](https://github.com/arenadata/gpdb/blob/7d470e595c7c77bcafeaf6a13482db43668fbca1/src/backend/access/aocs/aocsam.c#L859) right after [OpenAOCSDatumStreams()](https://github.com/arenadata/gpdb/blob/7d470e595c7c77bcafeaf6a13482db43668fbca1/src/backend/access/aocs/aocsam.c#L848)-> [InsertInitialAOCSFileSegInfo()](https://github.com/arenadata/gpdb/blob/7d470e595c7c77bcafeaf6a13482db43668fbca1/src/backend/access/aocs/aocsam.c#L772) and we get a positive `last_sequence` in result.
But `InsertInitialAOCSFileSegInfo()` is also called on master when we collect the fast sequence totals from segments on the end of transaction and we get `last_sequence = -1` in this case. Thanks that you have noticed this problem.

Reply:
I don't want to add  any redundant logic with `GetFastSequences()` and `InsertFastSequenceEntry()` functions in `InsertInitialAOCSFileSegInfo()`, so I have [changed](https://github.com/arenadata/gpdb/compare/7d470e595c7c77bcafeaf6a13482db43668fbca1..6245ee67a7ee8c2cca21ad1fd6bf1e5ee8a67af4) `gp_fastsequence` generation and modified one more test. Let's look at [Pivotal Concourse CI](https://github.com/greenplum-db/gpdb/pull/11371) after this force push - may be some additional tests fail.